synapse/push/httppusher.py

```diff
@@ -440,10 +496,41 @@ async def dispatch_push(
         rejected = []
         if "rejected" in resp:
             rejected = resp["rejected"]
-        if not rejected:
-            self.badge_count_last_call = badge
         return rejected
 
+    async def dispatch_push_event(
+        self,
+        event: EventBase,
+        tweaks: JsonMapping,
+        badge: int,
+    ) -> Union[bool, Iterable[str]]:
+        """Send a notification to the registered push gateway by building it
+        from an event.
+
+        Args:
+            event: the event
+            tweaks: tweaks to add into the `devices` section, and to decide the
+            priority to use
+            badge: unread count to send with the push notification
+
+        Returns:
+            False if an error occured when calling the push gateway, or an array of
+            rejected push keys otherwise. If this array is empty, the push fully
+            succeeded.
+        """
+        content, tweaks = await self._build_event_notification_content(
+            event, tweaks, badge
+        )
```

Comment:
I suspect this code would be simpler if we inlined `_build_event_notification_content` here? Or is it separate for testing or something?

I'm finding it quite difficult to follow the logic; I think because the tweaks go into a separate piece (under `devices`), but whether we include them or not depends on the internsls of `_build_event_notification_content`?